core/keystore/src/lib.rs

```diff
@@ -14,185 +14,180 @@
 // You should have received a copy of the GNU General Public License
 // along with Substrate. If not, see <http://www.gnu.org/licenses/>.
 
-//! Keystore (and session key management) for ed25519 based chains like Polkadot.
+//! Session key store is an in memory key store.
 
 #![warn(missing_docs)]
 
+use parking_lot::Mutex;
+use substrate_primitives::crypto::{KeyTypeId, Pair};
 use std::collections::HashMap;
-use std::path::PathBuf;
-use std::fs::{self, File};
-use std::io::{self, Write};
+use std::sync::Arc;
+use std::time::{Duration, Instant};
 
-use substrate_primitives::crypto::{KeyTypeId, Pair, Public};
+// 24h in seconds
+const INACTIVITY_PERIOD: u64 = 60 * 60 * 24;
 
 /// Keystore error.
 #[derive(Debug, derive_more::Display, derive_more::From)]
 pub enum Error {
-	/// IO error.
-	Io(io::Error),
-	/// JSON error.
-	Json(serde_json::Error),
-	/// Invalid password.
-	#[display(fmt="Invalid password")]
-	InvalidPassword,
-	/// Invalid BIP39 phrase
-	#[display(fmt="Invalid recovery phrase (BIP39) data")]
-	InvalidPhrase,
 	/// Invalid seed
 	#[display(fmt="Invalid seed")]
 	InvalidSeed,
 }
 
+impl std::error::Error for Error {}
+
 /// Keystore Result
 pub type Result<T> = std::result::Result<T, Error>;
 
-impl std::error::Error for Error {
-	fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
-		match self {
-			Error::Io(ref err) => Some(err),
-			Error::Json(ref err) => Some(err),
-			_ => None,
-		}
-	}
+/// Handler for `Store` events.
+trait StoreHandler: Send {
```

Comment:
Can we have a `const` value of keytype here and have `StoreHandler` `pair` arguement be properly typed?

Reply:
you mean a `type KeyType;` here to allow `fn on_add_session_key(&mut self, pair: KeyType)`?